presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java

```diff
@@ -120,6 +120,16 @@ private Expression rewriteExpression(Expression expression, Predicate<VariableRe
             subExpressions = filter(subExpressions, not(equalTo(expression)));
         }
 
+        if (!variableToExpressionPredicate(variableScope, types).apply(expression)) {
```

Comment:
Do we wanna have a session property to guard this?

Reply:
We can. Putting this PR up to start conversation on how exactly we want to go forward with it. If we want to go with this strict version we definitely should add a session property.